Add parser support for URLPattern API

jelee53 opened the pull request:
#### 781f0cd0ee913678fcca6360aea3b85d14a9f246
<pre>
Add parser support for URLPattern API
<a href="https://bugs.webkit.org/show_bug.cgi?id=282413">https://bugs.webkit.org/show_bug.cgi?id=282413</a>
<a href="https://rdar.apple.com/139032626">rdar://139032626</a>

Reviewed by Chris Dumez and Sihui Liu.

Implement parser support for URL Pattern API, see <a href="https://urlpattern.spec.whatwg.org/#pattern-parser">https://urlpattern.spec.whatwg.org/#pattern-parser</a>

* Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
(WebCore::canonicalizeOpaquePathname):
(WebCore::canonicalizePathname):
* Source/WebCore/Modules/url-pattern/URLPatternParser.cpp: Added.
(WebCore::URLPatternUtilities::URLPatternParser::URLPatternParser):
(WebCore::URLPatternUtilities::URLPatternParser::performParse):
(WebCore::URLPatternUtilities::URLPatternParser::tryToConsumeToken):
(WebCore::URLPatternUtilities::URLPatternParser::tryToConsumeRegexOrWildcardToken):
(WebCore::URLPatternUtilities::URLPatternParser::tryToConsumeModifierToken):
(WebCore::URLPatternUtilities::URLPatternParser::consumeText):
(WebCore::URLPatternUtilities::URLPatternParser::consumeRequiredToken):
(WebCore::URLPatternUtilities::URLPatternParser::maybeAddPartFromPendingFixedValue):
(WebCore::URLPatternUtilities::URLPatternParser::addPart):
(WebCore::URLPatternUtilities::URLPatternParser::isDuplicateName const):
(WebCore::URLPatternUtilities::parse):
(WebCore::URLPatternUtilities::generateSegmentWildcardRegexp):
(WebCore::URLPatternUtilities::escapeRegexStringForCharacters):
(WebCore::URLPatternUtilities::escapeRegexString):
* Source/WebCore/Modules/url-pattern/URLPatternParser.h: Added.
(WebCore::URLPatternUtilities::URLPatternParser::setTokenList):
(WebCore::URLPatternUtilities::URLPatternParser::getPartList):
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/286362@main">https://commits.webkit.org/286362@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/28b6d5f33ed84a76edfe23b653a5a0687f9afa7e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/75783 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/48/builds/54812 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/28635 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/80273 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/27049 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/77899 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/63954 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/3072 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/59430 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17597 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/78850 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/49/builds/63954 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/65096 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39789 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/49/builds/63954 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/22579 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/25376 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/49/builds/63954 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/22917 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/81744 "Built successfully") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/3133 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/1978 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/67657 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/3276 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/65063 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66962 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16684 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10913 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/9042 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/3081 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/5907 "Failed to build and analyze WebKit") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/3110 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/4045 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/3130 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternParser.cpp:
```diff
@@ -0,0 +1,341 @@
+    while (m_index < m_tokenList.size()) {
+        auto charToken = tryToConsumeToken(TokenType::Char);
+        auto nameToken = tryToConsumeToken(TokenType::Name);
+        auto regexOrWildcardToken =  tryToConsumeRegexOrWildcardToken(nameToken);
```
Too many spaces after the `=`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternParser.cpp:
```diff
@@ -0,0 +1,341 @@
+
+            auto modifierToken = tryToConsumeModifierToken();
+
+            maybeFunctionException = addPart(WTFMove(prefix), nameToken, regexOrWildcardToken, ""_s, modifierToken);
```
Why use `""_s` instead of `emptyString()`?

jelee53 replied:
I was implementing this part of the spec where suffix is the empty string. 
<img width="909" alt="Screenshot 2024-11-01 at 10 42 59 AM" src="https://github.com/user-attachments/assets/869f5c93-a509-4951-831f-9374eb22ad21">

Since suffix is not re-used at all, I wanted to transfer ownership of the suffix string to the part list. You can see in other parts, I do WTFMove(suffix). 

But emptyString() is const so I couldn't transfer ownership. So I used ""_s instead.

cdumez replied:
So use `const String&` instead of `String&&` for the parameter? As addressed elsewhere in my review feedback, you're passing a `String&&` around and end up calling `callEncodingCallback()` which doesn't make use of the r-value reference anyway (no internal WTFMove()).

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternParser.cpp:
```diff
@@ -0,0 +1,341 @@
+        if (fixedToken.isNull())
+            fixedToken = tryToConsumeToken(TokenType::EscapedChar);
+        else if (fixedToken.isInitialized()) {
+            m_pendingFixedValue = makeString(WTFMove(m_pendingFixedValue), fixedToken.value);
```
I recommend using std::exchange instead of `WTFMove()` here since `m_pendingFixedValue` is likely getting re-used after the move (which is not a great practice).

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternParser.cpp:
```diff
@@ -0,0 +1,341 @@
+        if (openToken.isInitialized()) {
+            String prefix = consumeText();
+            nameToken = tryToConsumeToken(TokenType::Name);
+            regexOrWildcardToken =  tryToConsumeRegexOrWildcardToken(nameToken);
```
too many spaces after =

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternParser.cpp:
```diff
@@ -0,0 +1,341 @@
+// https://urlpattern.spec.whatwg.org/#try-to-consume-a-token
+Token URLPatternParser::tryToConsumeToken(TokenType consumedType)
+{
+    ASSERT(m_index < m_tokenList.size());
```
This assertion is not helpful, WTF::Vector's operator[] already has such assertion (and a release assertion in fact).

jelee53 replied:
Hi Chris - this was a recommendation from the spec: 

<img width="524" alt="Screenshot 2024-11-01 at 10 34 57 AM" src="https://github.com/user-attachments/assets/9434bb1f-95e0-40f0-af93-f3424aa41820">

cdumez replied:
What's your point? We follow this recommendation already since operator[] does such assertion.

jelee53 replied:
I guess I was thinking it makes sense to catch the error at the caller level? Like why not catch the error before we actually try to index into token list with an invalid index?

And I wonder if that's what the original author was thinking? But I can change it to remove the assertion.

cdumez replied:
> I wonder if that's what the original author was thinking

The spec author knows nothing of our Vector::operator[] implementation.

> I guess I was thinking it makes sense to catch the error at the caller level? Like why not catch the error before we actually try to index into token list with an invalid index?

operator[] is inline, you're crashing one line earlier, that's it.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternParser.cpp:
```diff
@@ -0,0 +1,341 @@
+    if (nextToken.type != consumedType)
+        return { };
+
+    m_index = m_index + 1;
```
`++m_index;` ?

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternParser.cpp:
```diff
@@ -0,0 +1,341 @@
+    if (token.isInitialized())
+        return token;
+
+    token = tryToConsumeToken(TokenType::Asterisk);
```
`return tryToConsumeToken(TokenType::Asterisk);` ?

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternParser.cpp:
```diff
@@ -0,0 +1,341 @@
+String URLPatternParser::consumeText()
+{
+    StringBuilder result;
+    result.append(""_s);
```
It seems weird to append an empty string to a StringBuilder?

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternParser.cpp:
```diff
@@ -0,0 +1,341 @@
+    if (m_pendingFixedValue.isEmpty())
+        return { };
+
+    auto encodedValue = callEncodingCallback(m_callbackType, WTFMove(m_pendingFixedValue));
```
ditto about using `std::exchange()`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternParser.cpp:
```diff
@@ -0,0 +1,341 @@
+        regexValue = emptyString();
+    }
+
+    String name = emptyString();
```
I suspect all the assignments to `emptyString()` in this patch are not really correct, though this is not a big deal. I'm curious why we can't let the Strings be null.

jelee53 replied:
The spec says that it should be set to the empty string but you're right that they're not entirely necessary. Maybe Blink's implementation of strings is different. I will remove.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternParser.h:
```diff
@@ -0,0 +1,86 @@
+private:
+    EncodingCallbackType m_callbackType;
+    String m_segmentWildcardRegexp;
+    String m_pendingFixedValue { emptyString() };
```
Curious why we need to initialize to emptyString.

jelee53 replied:
I think it's the spec author's preference: 

See 
https://urlpattern.spec.whatwg.org/#pattern-parser
<img width="751" alt="Screenshot 2024-11-01 at 10 36 22 AM" src="https://github.com/user-attachments/assets/b797cc4d-890e-4134-9a08-3779d2cf300e">

https://urlpattern.spec.whatwg.org/#part
<img width="611" alt="Screenshot 2024-11-01 at 10 36 42 AM" src="https://github.com/user-attachments/assets/ddf0ae2b-6897-4bda-9024-fff54ea3d804">

I was also getting these kind of complaints from the windows build. 
```
C:\BW\Win-Build-EWS\build\Source\WebCore\Modules/url-pattern/URLPatternParser.cpp(191,130): error: missing field 'name' initializer [-Werror,-Wmissing-field-initializers]
  191 |     m_partList.append(Part { .type = PartType::FixedText, .value = encodedValue.releaseReturnValue(), .modifier = Modifier::None });
```

cdumez replied:
"I think it's the spec author's preference:" doesn't convince me. We want the behavior of the spec. The spec doesn't know our String is not empty by default.

cdumez replied:
With regards to the build error, it's standard C++. You're missing an initializer for `name` in your struct.
Something like:
```
struct Foo {
    String name { };
};
```

Note the `{ }` after the `name`.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternParser.h:
```diff
@@ -0,0 +1,86 @@
+    ExceptionOr<void> maybeAddPartFromPendingFixedValue();
+    ExceptionOr<void> addPart(String&& prefix, const Token& nameToken, const Token& regexpOrWildcardToken, String&& suffix, const Token& modifierToken);
+
+    bool isDuplicateName(StringView name);
```
can this be marked as `const` ?

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternParser.h:
```diff
@@ -0,0 +1,86 @@
+    size_t m_index { 0 };
+    int m_nextNumericName { 0 };
+
+    Token tryToConsumeToken(TokenType consumedType);
```
Member functions should come *before* data members (with an empty line in between).

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternParser.h:
```diff
@@ -0,0 +1,86 @@
+
+class URLPatternParser {
+public:
+    Vector<Token> m_tokenList;
```
We should not have public data members in a class.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternParser.h:
```diff
@@ -0,0 +1,86 @@
+class URLPatternParser {
+public:
+    Vector<Token> m_tokenList;
+    Vector<Part> m_partList;
```
We should not have public data members in a class.

szewai commented on Source/WebCore/Modules/url-pattern/URLPatternParser.h:
```diff
@@ -0,0 +1,90 @@
+    PartType type;
+    String value;
+    Modifier modifier;
+    String name { };
+    String prefix { };
+    String suffix { };
```
`{ }` is unnecessary; it should be null string by default -- Same for all the other places

cdumez replied:
What Sihui is saying is not necessarily true for POD structs. It depends on how the struct is being initialized. If you need `{ }` because your local compiler or one of the EWS tells you, this is fine.

szewai replied:
Yeah but String is not POD, so this struct cannot be POD?
I understand there could be warning about "missing initializer" if the caller does not initialize all members, but `Part` is constructed with designated initializers, which should invoke default constructor?
For the same reason, `{ }` seems unnecessary for String in `URLPatternStringOptions`.

cdumez replied:
I'll be clearer. This code is in the PR:
```
m_partList.append(Part { .type = PartType::FixedText, .value = encodedValue.releaseReturnValue(), .modifier = Modifier::None });
```
I believe this would not build on all compilers WITHOUT the *explicit* initializers that Jessica added. It may build on Mac but I'm fairly confident it won't build on some of the EWS.

So I disagree with the review suggestion. I think it will break EWS build. But feel free to try. As always, these explicit `{ }` in struct data members should only be added if actually needed to build. I think they are here but I could be wrong 🤷🏻

szewai replied:
ah I didn't know we have tried removing it before and ews bots failed. My point is the same: we should add it only when necessary. It seems wpe and gtk bots needed it based on previous test runs, then it's fine to keep it.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
```diff
@@ -216,9 +216,9 @@ ExceptionOr<String> processPathname(StringView pathnameValue, const StringView p
 
     if (WTF::URLParser::isSpecialScheme(protocolValue))
-        return canonicalizePathname(pathnameValue);
+        return canonicalizePathname(pathnameValue.toString());
```
If most call sites on canonicalizePathname() have to call `toString()` then likely `canonicalizePathname()` should keep taking in a `StringView`, rather than a `String&&`

jelee53 replied:
It needs to be String&& because otherwise, StringView may somehow release reference.

cdumez replied:
Sounds like you have a bug somewhere then.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
```diff
@@ -216,9 +216,9 @@ ExceptionOr<String> processPathname(StringView pathnameValue, const StringView p
+        return canonicalizePathname(pathnameValue.toString());
 
-    return canonicalizeOpaquePathname(pathnameValue);
+    return canonicalizeOpaquePathname(pathnameValue.toString());
```
ditto.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternParser.cpp:
```diff
@@ -0,0 +1,335 @@
+
+URLPatternParser::URLPatternParser(EncodingCallbackType type, String&& segmentWildcardRegexp)
+    : m_callbackType(type)
+    , m_segmentWildcardRegexp(segmentWildcardRegexp)
```
`WTFMove(segmentWildcardRegexp)`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternParser.cpp:
```diff
@@ -0,0 +1,335 @@
+            fixedToken = tryToConsumeToken(TokenType::EscapedChar);
+
+        if (!fixedToken.isNull()) {
+            std::exchange(m_pendingFixedValue, makeString(m_pendingFixedValue, fixedToken.value));
```
`m_pendingFixedValue = makeString(m_pendingFixedValue, fixedToken.value);` ?

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternParser.cpp:
```diff
@@ -0,0 +1,335 @@
+    }
+
+    if (nameToken.isNull() && regexpOrWildcardToken.isNull() && modifier == Modifier::None) {
+        std::exchange(m_pendingFixedValue, makeString(m_pendingFixedValue, prefix));
```
`m_pendingFixedValue  = makeString(m_pendingFixedValue, prefix);`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternParser.cpp:
```diff
@@ -0,0 +1,335 @@
+        name = nameToken.value.toString();
+    else if (regexpOrWildcardToken.isNull()) {
+        name = String::number(m_nextNumericName);
+        m_nextNumericName = m_nextNumericName + 1;
```
`++ m_nextNumericName;`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternParser.cpp:
```diff
@@ -0,0 +1,335 @@
+// https://urlpattern.spec.whatwg.org/#is-a-duplicate-name
+bool URLPatternParser::isDuplicateName(StringView name) const
+{
+    for (auto part : m_partList) {
```
`auto&` to avoid copying the part?

cdumez replied:
But also, this could be one line using `Vector::containsIf()` with a lambda.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternParser.h:
```diff
@@ -0,0 +1,90 @@
+};
+
+struct URLPatternStringOptions {
+    String delimiterCodepoint { emptyString() };
```
Does this really need `emptyString()`?

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternParser.h:
```diff
@@ -0,0 +1,90 @@
+    URLPatternParser(EncodingCallbackType, String&& segmentWildcardRegexp);
+    ExceptionOr<void> performParse(const URLPatternStringOptions& patternSettings);
+
+    void setTokenList(Vector<Token>&& tokenList) { m_tokenList = tokenList; }
```
`WTFMove(tokenList)`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternParser.h:
```diff
@@ -0,0 +1,90 @@
+    ExceptionOr<void> performParse(const URLPatternStringOptions& patternSettings);
+
+    void setTokenList(Vector<Token>&& tokenList) { m_tokenList = tokenList; }
+    Vector<Part> getPartList(void) { return m_partList; }
```
should return a `const Vector<Part>&` to not force the call site to copy. Also please omit `void` in the parameter list.

szewai replied:
Also a member function should marked const if possible:
`const Vector<Part> getPartList() const;`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternParser.h:
```diff
@@ -0,0 +1,90 @@
+    Vector<Part> m_partList;
+    EncodingCallbackType m_callbackType;
+    String m_segmentWildcardRegexp;
+    String m_pendingFixedValue { };
```
Please omit `{ }`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternParser.h:
```diff
@@ -0,0 +1,90 @@
+    int m_nextNumericName { 0 };
+};
+
+ExceptionOr<Vector<Part>> parse(StringView patternStringInput, const URLPatternStringOptions& patternSettings,  EncodingCallbackType);
```
Too many spaces before `EncodingCallbackType`

cdumez replied:
I think we can omit the parameter names.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternParser.h:
```diff
@@ -0,0 +1,90 @@
+ExceptionOr<Vector<Part>> parse(StringView patternStringInput, const URLPatternStringOptions& patternSettings,  EncodingCallbackType);
+
+// FIXME: Consider moving functions to somewhere generic, perhaps refactor Part to its own class.
+String generateSegmentWildcardRegexp(const URLPatternStringOptions& patternStringSettings);
```
I think we can omit `patternStringSettings`. It doesn't bring much info given the parameter type.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternParser.h:
```diff
@@ -0,0 +1,90 @@
+
+// FIXME: Consider moving functions to somewhere generic, perhaps refactor Part to its own class.
+String generateSegmentWildcardRegexp(const URLPatternStringOptions& patternStringSettings);
+String escapeRegexString(StringView input);
```
I think we can omit `input`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternParser.h:
```diff
@@ -0,0 +1,90 @@
+class URLPatternParser {
+public:
+    URLPatternParser(EncodingCallbackType, String&& segmentWildcardRegexp);
+    ExceptionOr<void> performParse(const URLPatternStringOptions& patternSettings);
```
I think we can omit the parameter name.

szewai commented on Source/WebCore/Modules/url-pattern/URLPatternParser.h:
```diff
@@ -0,0 +1,90 @@
+String generateSegmentWildcardRegexp(const URLPatternStringOptions& patternStringSettings);
+String escapeRegexString(StringView input);
+
+
```
extra blank line

szewai commented on Source/WebCore/Modules/url-pattern/URLPatternParser.h:
```diff
@@ -0,0 +1,90 @@
+
+private:
+    Token tryToConsumeToken(TokenType consumedType);
+    Token tryToConsumeRegexOrWildcardToken(const Token& nameToken);
```
The parameter names (`consumedType`, `nameToken`) are not needed

szewai commented on Source/WebCore/Modules/url-pattern/URLPatternParser.h:
```diff
@@ -0,0 +1,90 @@
+    Token tryToConsumeModifierToken();
+
+    String consumeText();
+    ExceptionOr<Token> consumeRequiredToken(TokenType requiredType);
```
parameter name seems unnecessary

szewai commented on Source/WebCore/Modules/url-pattern/URLPatternParser.cpp:
```diff
@@ -0,0 +1,335 @@
+        auto regexOrWildcardToken = tryToConsumeRegexOrWildcardToken(nameToken);
+
+        if (!nameToken.isNull() || !regexOrWildcardToken.isNull()) {
+            String prefix;
```
Spec says empty string, not null

cdumez replied:
The spec says empty string for all strings, I don't think we have to be too literal. I think what matters is the end behavior.

szewai commented on Source/WebCore/Modules/url-pattern/URLPatternParser.cpp:
```diff
@@ -0,0 +1,335 @@
+                prefix = charToken.value.toString();
+
+            if (!prefix.isEmpty() && prefix != patternSettings.prefixCodepoint)
+                std::exchange(m_pendingFixedValue, makeString(m_pendingFixedValue, std::exchange(prefix, { })));
```
Spec says sets prefix to empty string, not null.
Also it's more clear to write it as:
`m_pendingFixedValue = makeString(m_pendingFixedValue, std::exchange(prefix, emptyString());`

If we expect to keep appending to `m_pendingFixedValue`, it should probably be a `StringBuilder`.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternParser.cpp:
```diff
@@ -0,0 +1,334 @@
+// https://urlpattern.spec.whatwg.org/#try-to-consume-a-token
+Token URLPatternParser::tryToConsumeToken(TokenType type)
+{
+    auto nextToken = m_tokenList[m_index];
```
`auto&` to avoid a copy?

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternParser.cpp:
```diff
@@ -0,0 +1,334 @@
+// https://urlpattern.spec.whatwg.org/#is-a-duplicate-name
+bool URLPatternParser::isDuplicateName(StringView name) const
+{
+
```
Please remove extra blank line here.

szewai commented on Source/WebCore/Modules/url-pattern/URLPatternParser.h:
```diff
@@ -0,0 +1,89 @@
+
+struct URLPatternStringOptions {
+    String delimiterCodepoint { };
+    String prefixCodepoint { };
```
Please try removing `{ }`

jelee53 committed:
Add parser support for URLPattern API
https://bugs.webkit.org/show_bug.cgi?id=282413
rdar://139032626

Reviewed by Chris Dumez and Sihui Liu.

Implement parser support for URL Pattern API, see https://urlpattern.spec.whatwg.org/#pattern-parser

* Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
(WebCore::canonicalizeOpaquePathname):
(WebCore::canonicalizePathname):
* Source/WebCore/Modules/url-pattern/URLPatternParser.cpp: Added.
(WebCore::URLPatternUtilities::URLPatternParser::URLPatternParser):
(WebCore::URLPatternUtilities::URLPatternParser::performParse):
(WebCore::URLPatternUtilities::URLPatternParser::tryToConsumeToken):
(WebCore::URLPatternUtilities::URLPatternParser::tryToConsumeRegexOrWildcardToken):
(WebCore::URLPatternUtilities::URLPatternParser::tryToConsumeModifierToken):
(WebCore::URLPatternUtilities::URLPatternParser::consumeText):
(WebCore::URLPatternUtilities::URLPatternParser::consumeRequiredToken):
(WebCore::URLPatternUtilities::URLPatternParser::maybeAddPartFromPendingFixedValue):
(WebCore::URLPatternUtilities::URLPatternParser::addPart):
(WebCore::URLPatternUtilities::URLPatternParser::isDuplicateName const):
(WebCore::URLPatternUtilities::parse):
(WebCore::URLPatternUtilities::generateSegmentWildcardRegexp):
(WebCore::URLPatternUtilities::escapeRegexStringForCharacters):
(WebCore::URLPatternUtilities::escapeRegexString):
* Source/WebCore/Modules/url-pattern/URLPatternParser.h: Added.
(WebCore::URLPatternUtilities::URLPatternParser::setTokenList):
(WebCore::URLPatternUtilities::URLPatternParser::getPartList):
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/286362@main
